### ISSUE TITLE
Add events for user account and ban list add, change, and remove

### ISSUE THREAD

**Issue by tech10** [BearWare/TeamTalk5#842]
This would apply to server administrators.

At the moment, no clients are notified of any changes to user accounts or the ban list, which is a vulnerability if a malicious user is modifying such information, perhaps after gaining access to administrator rights in one manner or another. I propose that any connected client that's an administrator receive such events automatically. Whether or not the client displays those events to a user should be up to them. I would recommend they have an option between sound, text to speech, and a dialog box containing changes to account or ban information.

I have had experiences where a malicious user gained access to a server and removed accounts without other administrators being notified of such activity. If such information was presented, malicious users could be ban and removed far quicker, before an entire servers user or ban list is compromised unknowingly. I have written my own custom bot for such information, which continually pulls user account and ban information, comparing new information against the old. I have utilized this system to assist in ensuring authorized activity is occurring on a server, but if the server sent out such information to each administrator automatically, it would be far more efficient, and I believe, would help server administrators monitor unauthorized and authorized activity on their servers alike.

**Comment by bear101** [comment]
If you modify a user account then you manually have to kick all the account's users. Updating the user account of users who are already logged in would be a complicated and error prone process.

**Comment by tech10** [comment]
If you understood my issue and didn't misunderstand it, forgive my explanation below, but I do believe there is a vulnerability here that could be corrected if a simple change is made to the protocol. I want to be sure this is understood correctly, and if it remains closed, I won't say anything further on the topic.

This is not to modify an already logged in users account, that can easily be done already. This is to notify administrators of user accounts that have been changed by someone else, or a ban list change.

I think I know what you are referring to, but that isn't what I am referring to. If I surmise what I understand about what you stated, you thought I was referring to a logged in user account requiring that any updates to it be made immediately, altering the user rights of a connected user, changing their account status, and such an event taking effect immediately for them. This isn't what I was talking about, if that's what you were talking about, which would require that any logged in user of a changed account be kicked at current. That's an entirely different issue, and I agree with you, more complicated to implement at the moment.

I am referring to the server sending administrators an event whenever a user account or ban list change is made, so their client can receive such a notification and pass it along to them, to help improve the monitoring and security of a server. As I stated above, if someone malicious were to gain access to a server and remove all accounts, for instance, no other administrators would know about this, due to the fact that the server does not send other administrators any notification of this change in status. Users could remain connected for hours after this with no one the wiser until someone leaves and tries to reconnect, or another user can't connect to the server due to their account having been removed, not known by other administrators immediately because no events are sent out about such changes.

If this helps your understanding of the issue, think of it like this. Whenever a user connects and joins a channel, all other connected users are sent a notification of the event. My proposal is that whenever a user account or ban list update is made, whether that be added, changed, or removed, all other administrators will have that notification sent to their client. At the very least, they would be notified of the account name or ban list IP address that was changed by another administrator. Take the following example from the protocol, which could be expanded to include the ban list.

The client sends this to add a user account, and the server responds. The administrator account modifying accounts in this proposal has the username tech.

```
newaccount username="test" password="testing" usertype=1 userrights=259591 userdata=0 note="" initchan="" opchannels=[] audiocodeclimit=0 id=22
begin id=22
ok
end id=22
```

My proposal would then extend the protocol slightly, sending an event out to other administrators like this.

```
accountadd username="test" updatinguser="tech"
```

The same thing would then occur if an account is changed or removed, as would a notification then be sent out in such a manner to administrators for ban list updates in this proposal. This way, if a malicious user did start removing user accounts, administrators would know who it is immediately, could then ban them and remove their account, and take any other possible actions immediately, as they know about the changes being made.

I think this is a simpler change to implement than you might have thought, and if you knew exactly what I was referring to already, I apologize for the long explanation. I would respectfully ask that you consider this to help servers be more secure, I'm sure I'm not the only one that would appreciate such updates being sent to them to help protect their servers.
